packages/core/src/StateNode.ts

```diff
@@ -673,21 +673,19 @@ class StateNode<
     }
 
     const subStateKeys = keys(stateValue);
-    const subStateNodes: Array<
-      StateNode<TContext, any, TEvent, TTypestate>
-    > = subStateKeys.map((subStateKey) => this.getStateNode(subStateKey));
-
-    subStateNodes.push(this);
-
-    return subStateNodes.concat(
-      subStateKeys.reduce((allSubStateNodes, subStateKey) => {
-        const subStateNode = this.getStateNode(subStateKey).getStateNodes(
-          stateValue[subStateKey]
-        );
+    const subStateNodes: Array<StateNode<TContext, any, TEvent, TTypestate>> = [
+      this
+    ];
 
-        return allSubStateNodes.concat(subStateNode);
-      }, [] as Array<StateNode<TContext, any, TEvent, TTypestate>>)
+    subStateNodes.push(
+      ...flatten(
+        subStateKeys.map((subStateKey) =>
+          this.getStateNode(subStateKey).getStateNodes(stateValue[subStateKey])
+        )
+      )
     );
```

Comment:
it's an equivalent to the reduce on the left - I've found it harder to read this reduce with concat so I've just refactored this to push+map